internal/datastore/spanner/watch.go

```diff
@@ -164,3 +181,175 @@ func contextualizedCaveatFromValues(values map[string]any) (*core.Contextualized
 	}
 	return nil, nil
 }
+
+const maxSchemaWatchRetryCount = 15
+
+func (sd spannerDatastore) WatchSchema(ctx context.Context, afterRevision datastore.Revision) (<-chan *datastore.SchemaState, <-chan error) {
+	updates := make(chan *datastore.SchemaState, 10)
+	errs := make(chan error, 1)
+
+	lastRevision := afterRevision
+	retryCount := -1
+	go (func() {
+		running := true
+		for running {
+			retryCount++
+			retryHistogram.Observe(float64(retryCount))
+			if retryCount >= maxSchemaWatchRetryCount {
+				errs <- fmt.Errorf("schema watch terminated after %d retries", retryCount)
+				return
+			}
+
+			log.Debug().Str("revision", lastRevision.String()).Int("retry-count", retryCount).Msg("starting schema watch connection")
+
+			sd.watchSchemaWithoutRetry(ctx, lastRevision, func(update *datastore.SchemaState, err error) {
+				if err != nil {
+					// TODO(jschorr): If this is a terminal error, just quit instead of retrying.
+					running = true
+
+					// Sleep a bit for retrying.
+					pgxcommon.SleepOnErr(ctx, err, uint8(retryCount))
+					return
+				}
+
+				if update.Revision.GreaterThan(lastRevision) {
+					lastRevision = update.Revision
+				}
+
+				updates <- update
+				retryCount = -1
+			})
+		}
+	})()
+
+	return updates, errs
+}
+
+func (sd spannerDatastore) watchSchemaWithoutRetry(ctx context.Context, afterRevisionRaw datastore.Revision, processUpdate func(update *datastore.SchemaState, err error)) {
+	project, instance, database, err := parseDatabaseName(sd.database)
+	if err != nil {
+		processUpdate(nil, err)
+		return
+	}
+
+	afterRevision := afterRevisionRaw.(revision.Decimal)
+	reader, err := changestreams.NewReaderWithConfig(
+		ctx,
+		project,
+		instance,
+		database,
+		SchemaChangeStreamName,
+		changestreams.Config{
+			StartTimestamp:    timestampFromRevision(afterRevision),
+			HeartbeatInterval: 100 * time.Millisecond, // 100ms is the minimum allowed
```

Comment:
N spicedb replicas running 10 heartbeats per second could be a bunch of heartbeats. Do we want to be more conservative, or perhaps expose this in a parameter now?